### PR TITLE
Fixed mod interaction with SMAPI code

### PR DESCRIPTION
**Pull request by misternebula**
Creating a PR even though this has some limitations. Seeing as SMAPI also has these limitations, and it will take a while to rewrite this to fix that, I think this is okay for now.
APIs can only use common types. That is, types that are already available to OWML and both mods. That means no custom types.

### CODE REVIEW

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            // create field to store target instance
+            FieldBuilder targetField = proxyBuilder.DefineField("__Target", targetType, FieldAttributes.Private);
+
+            // create constructor which accepts target instance and sets field
+            {
+                ConstructorBuilder constructor = proxyBuilder.DefineConstructor(MethodAttributes.Public, CallingConventions.Standard | CallingConventions.HasThis, new[] { targetType });
+                ILGenerator il = constructor.GetILGenerator();
+
+                il.Emit(OpCodes.Ldarg_0); // this
+                // ReSharper disable once AssignNullToNotNullAttribute -- never null
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:43]: Resharper comment, remove

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+            // define proxy type
+            TypeBuilder proxyBuilder = moduleBuilder.DefineType(name, TypeAttributes.Public | TypeAttributes.Class);
+            proxyBuilder.AddInterfaceImplementation(interfaceType);
+
+            // create field to store target instance
+            FieldBuilder targetField = proxyBuilder.DefineField("__Target", targetType, FieldAttributes.Private);
+
+            // create constructor which accepts target instance and sets field
+            {
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:38]: This is more about my lack of C# knowledge, but any clue why this block starts here? I guess to separate the scope, but seems weird to me. Maybe it should be a separate method. Also, separating some of this stuff into individual methods would make for easier debugging and less comments needed.

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        private readonly Type ProxyType;
+
+        /// <summary>Construct an instance.</summary>
+        /// <param name="name">The type name to generate.</param>
+        /// <param name="moduleBuilder">The CLR module in which to create proxy classes.</param>
+        /// <param name="interfaceType">The interface type to implement.</param>
+        /// <param name="targetType">The target type.</param>
+        public InterfaceProxyBuilder(string name, ModuleBuilder moduleBuilder, Type interfaceType, Type targetType)
+        {
+            // validate
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:24]: unnecessary comment

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            if (name == null)
+                throw new ArgumentNullException(nameof(name));
+            if (targetType == null)
+                throw new ArgumentNullException(nameof(targetType));
+
+            // define proxy type
+            TypeBuilder proxyBuilder = moduleBuilder.DefineType(name, TypeAttributes.Public | TypeAttributes.Class);
+            proxyBuilder.AddInterfaceImplementation(interfaceType);
+
+            // create field to store target instance
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:34]: unnecessary comment

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        /// <param name="targetType">The target type.</param>
+        public InterfaceProxyBuilder(string name, ModuleBuilder moduleBuilder, Type interfaceType, Type targetType)
+        {
+            // validate
+            if (name == null)
+                throw new ArgumentNullException(nameof(name));
+            if (targetType == null)
+                throw new ArgumentNullException(nameof(targetType));
+
+            // define proxy type
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:30]: unnecessary comment

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            foreach (MethodInfo proxyMethod in interfaceType.GetMethods())
+            {
+                var targetMethod = targetType.GetMethod(proxyMethod.Name, proxyMethod.GetParameters().Select(a => a.ParameterType).ToArray());
+                if (targetMethod == null)
+                    throw new InvalidOperationException($"The {interfaceType.FullName} interface defines method {proxyMethod.Name} which doesn't exist in the API.");
+
+                this.ProxyMethod(proxyBuilder, targetMethod, targetField);
+            }
+
+            // save info
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:61]: unnecessary comment

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                il.Emit(OpCodes.Ldarg_0); // this
+                // ReSharper disable once AssignNullToNotNullAttribute -- never null
+                il.Emit(OpCodes.Call, typeof(object).GetConstructor(new Type[0])); // call base constructor
+                il.Emit(OpCodes.Ldarg_0);      // this
+                il.Emit(OpCodes.Ldarg_1);      // load argument
+                il.Emit(OpCodes.Stfld, targetField); // set field to loaded argument
+                il.Emit(OpCodes.Ret);
+            }
+
+            // proxy methods
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:51]: unnecessary comment

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            Type[] argTypes = target.GetParameters().Select(a => a.ParameterType).ToArray();
+
+            // create method
+            MethodBuilder methodBuilder = proxyBuilder.DefineMethod(target.Name, MethodAttributes.Public | MethodAttributes.Final | MethodAttributes.Virtual);
+            methodBuilder.SetParameters(argTypes);
+            methodBuilder.SetReturnType(target.ReturnType);
+
+            // create method body
+            {
```
Comment (Raicuparta) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:91]: OK I really don't know what these blocks are for 😵

File: OWML.Common/IModInteraction.cs [hunk 9]
```diff
@@ -8,7 +8,7 @@ public interface IModInteraction
         IList<IModBehaviour> GetDependants(string dependencyUniqueName);
         IList<IModBehaviour> GetDependencies(string uniqueName);
         IModBehaviour GetMod(string uniqueName);
-        T GetMod<T>(string uniqueName) where T : IModBehaviour;
+        T GetApi<T>(string uniqueName) where T : class;
```
Comment (amazingalek) [inline OWML.Common/IModInteraction.cs:11]: Sure about calling it Api instead of Mod?

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Text;
+
+namespace OWML.ModHelper.Interaction
+{
+    internal class InterfaceProxyBuilder
```
Comment (amazingalek) [inline OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:10]: Let's clearly mark this file as based on SMAPI code :)
F.ex. you can link to this page: https://github.com/Pathoschild/SMAPI/tree/c4a82418ac8b09a6965052f5c9173928457fba52/src/SMAPI/Framework/Reflection

File: OWML.ModHelper.Interaction/InterfaceProxyFactory.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Text;
+
+namespace OWML.ModHelper.Interaction
+{
+    public class InterfaceProxyFactory
```
Comment (amazingalek) [inline OWML.ModHelper.Interaction/InterfaceProxyFactory.cs:10]: Same as above, I'd like it to be 100% clear this is based on SMAPI code :)

File: OWML.ModHelper/ModBehaviour.cs [hunk 12]
```diff
@@ -28,5 +31,7 @@ public IList<IModBehaviour> GetDependencies()
         {
             return ModHelper.Interaction.GetDependencies(ModHelper.Manifest.UniqueName);
         }
+
+        public virtual object GetApi() => null;
```
Comment (amazingalek) [inline OWML.ModHelper/ModBehaviour.cs:35]: Huh? :P

Comment (misternebula) [reply OWML.ModHelper/ModBehaviour.cs:35]: this gets called to get the mod API. a mod would override that method to return the API class

File: OWML.Common/IModInteraction.cs (hunk 9 above)
Comment (misternebula) [reply OWML.Common/IModInteraction.cs:11]: Well it does return the API of the mod... so maybe GetModAPI<T>() ?

File: OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs (hunk 10 above)
Comment (misternebula) [reply OWML.ModHelper.Interaction/InterfaceProxyBuilder.cs:10]: done!

File: OWML.ModHelper.Interaction/InterfaceProxyFactory.cs (hunk 11 above)
Comment (misternebula) [reply OWML.ModHelper.Interaction/InterfaceProxyFactory.cs:10]: done!

File: OWML.ModHelper/ModBehaviour.cs (hunk 12 above)
Comment (misternebula) [reply OWML.ModHelper/ModBehaviour.cs:35]: basically the same as the Configure method, except it returns the interface

Comment (amazingalek) [reply OWML.ModHelper/ModBehaviour.cs:35]: Ah, cool. Could you make it look like Configure (empty method with curly brackets), and put them together? Or doesn't it work then?

File: OWML.ModHelper.Interaction/ModInteraction.cs [hunk 13]
```diff
@@ -64,10 +72,26 @@ public IModBehaviour GetMod(string uniqueName)
...
+            {
+                return null;
+            }
+
+            if (inter is TInterface castInter)
+            {
+                return castInter;
+            }
+
+            return _proxyFactory.CreateProxy<TInterface>(inter, _modManifest.UniqueName, uniqueName);
```
Comment (amazingalek) [inline OWML.ModHelper.Interaction/ModInteraction.cs:94]: What's the difference between _modManifest.UniqueName and uniqueName? If they're the same, the manifest could be removed?

File: OWML.Common/IModInteraction.cs (hunk 9 above)
Comment (amazingalek) [reply OWML.Common/IModInteraction.cs]: Good idea :)

File: OWML.ModHelper/ModBehaviour.cs (hunk 12 above)
Comment (misternebula) [reply OWML.ModHelper/ModBehaviour.cs:35]: the method needs to returns something, so I guess I could just put a return null in the brackets to make it more like configure

File: OWML.ModHelper.Interaction/ModInteraction.cs (hunk 13 above)
Comment (misternebula) [reply OWML.ModHelper.Interaction/ModInteraction.cs:94]: ones for the mod using the api, ones for the mod providing the api (i think)

Comment (misternebula) [reply OWML.ModHelper.Interaction/ModInteraction.cs:94]: probably should rename uniqueName

File: OWML.ModHelper/ModBehaviour.cs (hunk 12 above)
Comment (amazingalek) [reply OWML.ModHelper/ModBehaviour.cs:35]: Ah, whatever. Just put the virtual stuff together :)

File: OWML.ModHelper.Interaction/ModInteraction.cs (hunk 13 above)
Comment (amazingalek) [reply OWML.ModHelper.Interaction/ModInteraction.cs:94]: Hm, why would both be needed? 🤔 for logging/errors?

**Pull request merged**